Core/GameObject: Change some trap stuff.

Golrag opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->
**Notes:**

I don't know if this is how it's supposed to work. Might be wrong/not blizzlike.
It's an old commit, might not be up to standards.

**Changes proposed:**

-  Fix traps that are triggered automatically while they shouldn't
-  Make previous traps (example: buffs in battlegrounds) depend on their areatriggers
   - The following battlegrounds have been changes:
      - Warsong Gulch
      - Eye of the Storm
      - Arathi Bassin
-  Add script for Massive Seaforium Charge

**Target branch(es):** 

- 3.3.5

**Issues addressed:**

Closes:
- #6626 (not everything has been fixed i think (mount example))
- #4590 only the trap detection part (did not check waypoints etc)

**Tests performed:**

- Builds
- Went to ICC, waited a couple of seconds. Nothing happend (good). Walked over a trap. It triggered.
- Went to the changed battlegrounds. Areatriggers that trigger work. Those that don't trigger (with .debug areatrigger) don't work.


**Known issues and TODO list:**

- [x] Define what data2 is. On master branch it is radius, on 3.3.5 it is diameter (same values in both databases)
- [x] Once this is defined, check the range on hunter traps etc. On retail they use areatriggers for that now.
    - The seem to be talking about radius. see: https://wow.gamepedia.com/Freezing_Trap
- [x] Define what data4 is. On master it is charges, on 3.3.5 it is type (same values in both databases)
- [x] Fel Reaver in Eye of the Storm is not working (areatrigger doesn't trigger)
- [x] Blacksmith in Arathi Bassin is not working (areatrigger doesn't trigger)
- [ ] Rogue disarm ? Maybe out of scope.

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Golrag committed:
Core/GameObject: Fix some traps:

* Fix traps that triggered automatically (e.g. Spirit Alarm).
* Fix Massive Seaforium Explosive:
    - Fix disarm
    - Explodes after 10 seconds
* Allow Battleground Buffs to work with areatriggers in the following battelgrounds:
    - Warsong Gulch
    - Eye of the Storm
    - Arathi Bassin
* Renamed diameter to radius
* Traps now use exact distance with their radius instead of including their geobox.

Warpten commented on src/server/game/Entities/GameObject/GameObject.cpp:
```diff
@@ -1523,7 +1489,7 @@ void GameObject::Use(Unit* user)
             m_cooldownTime = GameTime::GetGameTimeMS() + (goInfo->trap.cooldown ? goInfo->trap.cooldown : uint32(4)) * IN_MILLISECONDS;   // template or 4 seconds
 
-            if (goInfo->trap.type == 1)         // Deactivate after trigger
+            if (goInfo->trap.type)         // Deactivate after trigger
```
Explicit !=0, this is not boolish

Golrag commented:
Oh, I found this comment & piece of code in ```BattlegroundEY.cpp``` that might explain why the CMSG_AREATRIGGER isn't called.
```cpp
/// @workaround The original AreaTrigger is covered by a bigger one and not triggered on client side.
if (point == FEL_REAVER && m_PointOwnedByTeam[point] == player->GetTeam())
    if (m_FlagState && GetFlagPickerGUID() == player->GetGUID())
        if (player->GetDistance(2044.0f, 1729.729f, 1190.03f) < 3.0f)
            EventPlayerCapturedFlag(player, BG_EY_OBJECT_FLAG_FEL_REAVER);
```

Golrag commented:
This is really a bad way to test it but it seems to me that "diameter" should be radius. see https://puu.sh/Gz0Ug/ab11320948.mp4

No sniffs, so... can't  confirm a 100%. Might need to include object size? idk.

Golrag committed:
Remove SetLootState from script

Golrag committed:
Implement trap charges. This is completely guessed behavior at best :)

- The traps in ICC despawn 5 seconds after triggering, which is equal to their cooldown

Shauren commented on src/server/game/Battlegrounds/Zones/BattlegroundAB.cpp:
```diff
@@ -183,6 +186,16 @@ void BattlegroundAB::PostUpdateImpl(uint32 diff)
             EndBattleground(HORDE);
+
+        /// @workaround The original AreaTrigger is covered by a bigger one and not triggered on client side.
+        for (BattlegroundPlayerMap::const_iterator itr = GetPlayers().begin(); itr != GetPlayers().end(); ++itr)
```
Yeaaaaah, don't do this on every update like this

Golrag replied:
Hmmm.. What would you suggest as an alternative?

Golrag replied:
Is this something we could do in ```UpdatePositionData``` ? Could add all related area triggers there.

Shauren commented on src/server/game/Entities/GameObject/GameObject.cpp:
```diff
@@ -624,47 +619,16 @@ void GameObject::Update(uint32 diff)
                     // Pointer to appropriate target if found any
                     Unit* target = nullptr;
 
-                    /// @todo this hack with search required until GO casting not implemented
```
This is a wrong change. Thanks to a new field in GameObjectTemplate in master branch, we know that traps permanently spawned only trigger for players and summoned traps for any unit

Golrag replied:
Hmmm, the spirit alarms in ICC also trigger on hunter pets and eye of killrog. They might be summoned traps tho.
I noticed that while running around i didnt trigger a trap but then later at that same spot there was one.

So... Ignore that spirit alarm case and implement the check like it was before?

Shauren commented on src/server/game/Entities/GameObject/GameObject.cpp:
```diff
@@ -722,35 +686,34 @@ void GameObject::Update(uint32 diff)
+                    m_cooldownTime = GameTime::GetGameTimeMS() + (goInfo->trap.cooldown ? goInfo->trap.cooldown : uint32(4)) * IN_MILLISECONDS;
+
+                    bool shouldDeactivateImmediatly = goInfo->GetCharges() > 0 && goInfo->trap.cooldown == 0;
+                    shouldDeactivateImmediatly = shouldDeactivateImmediatly || GetUseCount() == goInfo->GetCharges();
```
eeeeeeeeeeeeeeeeh, couldn't you have written that as one expression?

Shauren commented on src/server/game/Entities/GameObject/GameObjectData.h:
```diff
@@ -122,9 +122,9 @@ struct GameObjectTemplate
             uint32 lockId;                                  //0 -> Lock.dbc
             uint32 level;                                   //1
-            uint32 diameter;                                //2 diameter for trap activation
+            uint32 radius;                                  //2 radius for trap activation
```
The name in 3.3.5 client is diameter but is used like radius (fix use, don't rename)

Golrag replied:
Well.. thats.. confusing. Thanks

Shauren commented on src/server/game/Entities/GameObject/GameObjectData.h:
```diff
@@ -122,9 +122,9 @@ struct GameObjectTemplate
+            uint32 radius;                                  //2 radius for trap activation
             uint32 spellId;                                 //3
-            uint32 type;                                    //4 0 trap with no despawn after cast. 1 trap despawns after cast. 2 bomb casts on spawn.
+            uint32 charges;                                 //4 0 trap with no despawn after cast. 1 trap despawns after cast. 2 trap despawns after cooldown after cast.
```
Make up your mind here... you renamed it to `charges` and also at the same time adjusted the comment as if it was `type`

Golrag replied:
Well.. I don't know to be honest. Now it will check increase the use counter then check if its equal to the charges value, if so, despawn. To have that behavior of the spirit alarms (despawning after 5 seconds / cooldown).

Golrag committed:
Changes based on feedback

Golrag commented:
There is also ```GameObject::Use(Unit* user)```. Not sure what the correct behavior should be. Trigger the trap? Disarm it? Set lootstate on `READY`?

HoustonCore commented:
@Golrag any news?)

Golrag commented:
Some things still fall under the "i dont know what is blizzlike"-category. Like the charges/type thing.
Then we have the AreaTrigger issues where i shouldnt look up every update tick, might do it every 500ms if thats acceptable, havent given it much thought yet.

offl commented:
Splitting this into 2-3 separate PRs(especially BG stuff and NYI areattriggers and scripts) can help to review and merge this faster

Golrag commented:
I could take a look at that

jackpoz commented:
Closing this PR as it seems abandoned. Feel free to ask here to reopen it if anyone would like to work on it.
